### PR TITLE
API method to list all commits of a repository

### PR DESCRIPTION
**Pull request by Mikescher**
This adds a new API Endpoint to list the commits of a repository:  
`/api/v1/repos/{user}/{repo}/git/commits?sha={id}&page={p}`
([equivalent github api](https://developer.github.com/v3/repos/commits/#list-commits-on-a-repository))

This functionality was a few times requested ( #3710 ; #5181 ) but I think nobody ever implemented it and because I need it for a personal project I tried my hands on it :)

I'm not really proficient in go, but I hope I did everything ok.

### CODE REVIEW

File: routers/api/v1/repo/commits.go
Comment (lunny) [inline routers/api/v1/repo/commits.go]: Why not merge the two `for commitPointer != nil {`?

**Review by lunny — CHANGES_REQUESTED** [review]

File: integrations/api_repo_git_commits_test.go
Comment (lunny) [inline integrations/api_repo_git_commits_test.go]: Also assert the JSON result is better.

**Review by lunny — CHANGES_REQUESTED** [review]

**Comment by mmarif4u** [comment]
I was going to open an issue for this, but thanks to this new PR.

Few questions @Mikescher,

- what is git here? is this part of the path?
- how many commits will it return by default or per page?

Minor feature request, is it possible to add boolean(true/false) flag to return in json for pagination if there is more than one page. That will make it more convinient instead of calling the end point for more results when there are not.

Thanks for the work.

File: routers/api/v1/repo/commits.go
Comment (Mikescher) [reply routers/api/v1/repo/commits.go]: Because I didn't know List had a .Len() method ^^
I changed it now :)

File: integrations/api_repo_git_commits_test.go
Comment (Mikescher) [reply integrations/api_repo_git_commits_test.go]: Okay, I also added a few more test cases (requesting an empty page, requesting a different branch)

**Comment by Mikescher** [comment]
> what is git here? is this part of the path?

Well, the `repos/{user}/{repo}/git/<sha>` method did already exist (to get a single commit) so I thought I just keep the same schema

> how many commits will it return by default or per page?

The page size is the one specified in `git.CommitsRangeSize`, currently it's 50 [1]

> Is it possible to add boolean(true/false) flag to return in json for pagination if there is more than one page. That will make it more convinient instead of calling the end point for more results when there are not.

Hmm, I kinda liked it that it is (more or less) the same format as the Github API.
Perhaps I could add two header fields, `X-CurrentPage` and `X-PageCount`?


[1] https://github.com/go-gitea/gitea/blob/master/vendor/code.gitea.io/git/repo_commit.go#L229

**Comment by mmarif4u** [comment]
Thanks @Mikescher.

I was more into stackoverflow api.
`"has_more": true`

https://api.stackexchange.com/docs/questions#order=desc&sort=activity&filter=default&site=stackoverflow&run=true

**Comment by Mikescher** [comment]
@ mmarif4u
Yeah, that would also be a possibility.  
There are currently a few other API method in gitea that return a paginated list:
 - `/repos/{owner}/{repo}/issuesList` (List a repositories issues) 
 - `/repos/{owner}/{repo}/pulls` (List a repo's pull requests) 

and they all follows the same design (directly return an array).

Would it be really be a problem for you if I put the `has_more` information into an response_header field? I kinda like the idea :)

**Comment by mmarif4u** [comment]
Anything will work with me as i need this for the Android app to implement, so i no need to call the API the 2nd time if there are no more records for the call.
https://gitlab.com/mmarif4u/gitnex

**Comment by Mikescher** [comment]
@mmarif4u:
Hey your the gitnex guy, I'm actually using your app on my phone :D

I added a few Response headers: `X-Page`, `X-HasMore`, `X-PageCount`, `X-PerPage` and `X-Total`.  
And the I spend a stupid amount of time to get swagger to recognize them...

I hope they are enough for your use case.

**Comment by Mikescher** [comment]
*(rebased pull request cause of #6364)*

Imo everything is cleared up and the PR can be reviewed

**Comment by Mikescher** [comment]
Well, I have no idea why the drone CI build suddenly fails. The only thing I did was rebasing my commits because there where quite a lot of commits since I created this PR...  

It look like the `go vendor` build step is downloading the latest versions of everything in /vendor/ and is then failing because it's different from the used versions - but no idea why it's suddenly doing that...

Also: Is there something I should change to get  this merged or just wait until someone has time to look at it?

**Comment by zeripath** [comment]
Hmm... Why can't you just use git for this?

**Comment by lunny** [comment]
@zeripath for an andorid client or others like, it's meaningful.

**Comment by Mikescher** [comment]
Just for everyone's information: I will stop updating/maintaining this PR.

I wrote that patch 2 months ago and kinda had the hope that it would either get merged or closed. And for the last two months I had to update this PR every time something in the project changed so that it couldn't be auto-merged anymore or so that the CI fails...  
I hoped (because it *had* the milestone 1.9.0) that this would end in June now - but I guess not.

So, no hard feelings - if you want you can close this PR and if someone wants to do something like this later feel free to use/reuse my code here.

Btw: I had to solve my initial problem by writing a separate program that accesses the raw gitea git-repositories on the filesystem and returns the commits when requested.

**Comment by lunny** [comment]
@Mikescher Thanks for your effort on this PR. And sorry for our slow review process. There are so many PRs and issues needed for maintainers to focus and resolve. But I think the maintainers will still spend time on reviewing this PR and help resolve conflicts.

**Comment by strahe** [comment]
This is really a useful feature, hope to include it as soon as possible. thanks.

**Comment by richmahn** [comment]
This would be good to have this. I would review it if the conflicts can be resolved, @Mikescher. I do hope the commits API endpoints can get developed!

**Comment by Mikescher** [comment]
> 
> 
> This would be good to have this. I would review it if the conflicts can be resolved, @Mikescher. I do hope the commits API endpoints can get developed!

Okay, I updated my branch.

File: routers/api/v1/repo/commits.go
Comment (sapk) [inline routers/api/v1/repo/commits.go]: This part would be better under [/routers/api/v1/swagger](https://github.com/go-gitea/gitea/tree/master/routers/api/v1/swagger)

File: routers/api/v1/repo/commits.go
Comment (sapk) [inline routers/api/v1/repo/commits.go]: This could be a sub-function similar to https://github.com/go-gitea/gitea/blob/37b6ee178705cd7e545ca3a8ae00b1f6cee89083/routers/api/v1/convert/convert.go#L50 That cloud be left in this file not necessarily to be put in convert module.
Furthermore, maybe de-duplicate with some parts with GetSingleCommit func.

**Review by sapk — COMMENTED** [review]
Seems pretty good. Thanks for the contrib and sorry for late review.
This PR could also be usefull for async (js) display of commits history. 
Some parts could be de-duplicate with GetSingleCommit and other should be better in other folder but it is ok like that and those could be done later.
I will approve later after testing it.

File: routers/api/v1/repo/commits.go
Comment (richmahn) [inline routers/api/v1/repo/commits.go]: The string part of Trees.URL should be `"/git/trees/"` (otherwise is broken link). I know you took it from the getSingleCommit function probably, and that was wrong. Was just fixed in a PR to both develop and release/v1.9

File: routers/api/v1/repo/commits.go
Comment (richmahn) [inline routers/api/v1/repo/commits.go]: The string part of HTMLURL should be `"/commit/"` (no "s"). I know you took it from the getSingleCommit function probably, and that was wrong. Was just fixed in a PR to both develop and release/v1.9. The URL here is valid, but doesn't give you the commit's diff, just a list of commits.

**Review by richmahn — CHANGES_REQUESTED** [review]

File: routers/api/v1/repo/commits.go
Comment (richmahn) [inline routers/api/v1/repo/commits.go]: This should be something like `"http://localhost:3000/api/v1/repos/root/test/git/commits/3f312aa93b50477c0d23e1c27b2563d4025f466d",` yet I'm getting it without a SHA: `"http://localhost:3000/api/v1/repos/root/test/git/commits/",`

**Review by richmahn — CHANGES_REQUESTED** [review]

**Comment by richmahn** [comment]
@Mikescher See reviews above.

**Review by richmahn — CHANGES_REQUESTED** [review]
Ok, found an error where if I make an empty repo (not initialized with files), http://localhost:3000/api/v1/repos/root/empty/git/commits gives a 500 error with HTML, but https://api.github.com/repos/richmahn/empty/commits gives a much nicer JSON error message that the "Git Repository is empty".

File: routers/api/v1/api.go
Comment (richmahn) [inline routers/api/v1/api.go]: GitHub actually does this as `/repos/:owner/:repo/commits`, and doesn't have a listing for `GET /repos/:owner/:repo/git/commits`. Maybe we should too? (https://developer.github.com/v3/repos/commits/#list-commits-on-a-repository)

Comment (lunny) [reply routers/api/v1/api.go]: I think we can do a broken change on v1.10 to follow github's API.

Comment (richmahn) [reply routers/api/v1/api.go]: @Mikescher Hopefully this route path can be changed. I believe there is a placeholder for `:owner/:repo/commits` as we also need the single commit for the non-git route, as that one is to show the files that have been changed (actually have developers that would like this so their app can quickly see if anything changed since last commit and what without having a git client and without having to clone a repo), so if you can put your m.Get("", ...) call there, we can then get one that takes a SHA as well.

File: routers/api/v1/api.go
Comment (lunny) [inline routers/api/v1/api.go]: Since next line has `reqRepoReader(models.UnitTypeCode)`, this line's is unnecessary.

Comment (Mikescher) [reply routers/api/v1/api.go]: Ok, I removed it

File: routers/api/v1/api.go
Comment (Mikescher) [reply routers/api/v1/api.go]: Okay I changed the api path from "/repos/{owner}/{repo}/git/commits" to "/repos/{owner}/{repo}/commits".  
I think that's what you meant?

File: routers/api/v1/repo/commits.go
Comment (Mikescher) [reply routers/api/v1/repo/commits.go]: okay, moved

File: routers/api/v1/repo/commits.go
Comment (Mikescher) [reply routers/api/v1/repo/commits.go]: fixed

File: routers/api/v1/repo/commits.go
Comment (Mikescher) [reply routers/api/v1/repo/commits.go]: fixed

File: routers/api/v1/repo/commits.go
Comment (Mikescher) [reply routers/api/v1/repo/commits.go]: fixed

File: routers/api/v1/api.go
Comment (richmahn) [reply routers/api/v1/api.go]: @Mikescher yeah, perfect!!

**Comment by richmahn** [comment]
@Mikescher Now just have the request to fix the 500 error you get if the repo you're querying the commits for is empty.

File: routers/api/v1/repo/commits.go
Comment (lafriks) [inline routers/api/v1/repo/commits.go]: Why remove return??

**Review by lafriks — CHANGES_REQUESTED** [review]

File: routers/api/v1/repo/commits.go
Comment (Mikescher) [reply routers/api/v1/repo/commits.go]: accident. fixed now.

File: routers/api/v1/repo/commits.go
Comment (lafriks) [inline routers/api/v1/repo/commits.go]:
```suggestion
		ctx.JSON(409, api.APIError{
			Message: "Git Repository is empty.",
			URL: setting.API.SwaggerURL,
		}
```

File: routers/api/v1/repo/commits.go
Comment (lafriks) [inline routers/api/v1/repo/commits.go]:
```suggestion
```

**Review by richmahn — APPROVED** [review]
👍

**Comment by lunny** [comment]
LGTM except what @sapk said https://github.com/go-gitea/gitea/pull/6408/files#r306415623

**Comment by sapk** [comment]
@Mikescher Can you make the suggested refactor?
@lunny @lafriks if not possible are you ok to merge as this and I will send the refactor after merge?

**Comment by Mikescher** [comment]
> @Mikescher Can you make the suggested refactor?

Okay I moved some code to a commit `toCommit` method, I hope that was what you meant?

I'm currently in the middle of my exams so I don't have that much free time...

**Comment by sapk** [comment]
That exactly what I mean and I totally understand that you may have no time for that so that why I suggest to merge as this. But this is perfect. Thanks again for your time.

**Review by sapk — APPROVED** [review]

**Review by lafriks — APPROVED** [review]

**Comment by sapk** [comment]
Looking at last comment from @lunny, I think I can merge this.

Thanks again for your contrib @Mikescher

**Comment by 24unix** [comment]
Is there any kind of special configuration needed to get the headers?
For me it works (on a private repo) with
`curl -X GET -I "https://git.xxx.net/api/v1/repos/xxx.net/yyy/commits" -H "accept: application/json" -H "Authorization: token scrambledtoken"
` 
but with an axios request I just receive the content-type in the header. I spent hours finding the mistake by myself but to no avail :( found some things related to cors stuff, but it is enabled in my app.ini.
So, what do I have to change in app.ini to get those headers with axios? Or, even better, would it be possible to return X-Total in data, like lenght?

**Comment by zeripath** [comment]
The headers required are: "accept: application/json" and "Authorization: token scrambledtoken"

**Comment by 24unix** [comment]
Thanks for the very fast reply, but it didn't help :-(
I guess it's something stupid I don't see …

screen

`axios.get(`https://git.xxx.net/api/v1/repos/xxxx.net/yyy/commits?page=${this.currentPage}&limit=5`,
					{
						headers: {
							Authorization: 'token ' + "bfebd3123c259f249d1292e2a3ce6234abe46f1a",
						    	accept: "application/json"
						}
					})
				.then((response) => {`

I know that the limit is unused, it's copy&paste from another view, will look for a limit solution later …

**Comment by zeripath** [comment]
Is Axios actually sending a get? Do you have access to the server router logs - just to double check that it's querying with GET. It could be useful to see what Axios is sending by checking the browser network logs.

Seeing that the curl command works implies that the problem is at the Axios end.

**Comment by 24unix** [comment]
This are the logs:
`176.94.8.122 - - [31/Jul/2020:21:39:01 +0000] "OPTIONS /api/v1/repos/xxx.net/yyy/commits?page=1&limit=5 HTTP/1.1" 200 4363 "https://127.0.0.1:8000/" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_16_0) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/84.0.4147.105 Safari/537.36"
176.94.8.122 - - [31/Jul/2020:21:39:01 +0000] "GET /api/v1/repos/xxx.net/yyy/commits?page=1&limit=5 HTTP/1.1" 200 8547 "https://127.0.0.1:8000/" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_16_0) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/84.0.4147.105 Safari/537.36"
`

**Comment by zeripath** [comment]
Axios is sending an OPTIONS request not a GET

**Comment by 24unix** [comment]
Oh, yes, overlooked it lasts night. Found that:
https://github.com/axios/axios/issues/475
It's a preflight check from axios to check capabilities. But axios still won't work.

**Comment by 24unix** [comment]
Nobody here using axios with gitea? Or am I the only one with that problem?

**Comment by 6543** [comment]
@24unix looks like not

and if you have more issues pleace create a new issue and just reverence merged pulls :)
this keep unrelated stuff seperated :)

**Pull request merged**
